### PR TITLE
Select output reference tensor correctly for pointwise scheduler

### PR DESCRIPTION
**Pull request by rdspring1**
* Pointwise scheduler heuristics requires a minimum number of axes.
  * The output reference tensor should respect this requirement.

```
def t(input1, input2, scalar):
  _unsafe_view = torch.ops.aten._unsafe_view(input1, [8, 512, 3072]);  mm = None
  add_ = torch.ops.aten.add_(_unsafe_view, input2);  _unsafe_view = primals_1 = None
  gelu_ = torch.ops.aten.gelu(add_)
  view_ = torch.ops.aten.view(gelu_, [4096, 3072])
  mul_ = torch.ops.aten.mul(add_, scalar);  add_ = _tensor_constant2 = None
  return [view_, mul_]
```

In the failing case, there are two output tensors that can become the reference tensor. Both tensors map to all iterDomains in the fusion. However, the `break_point` selected by the pointwise heuristics is 2, so the correct reference tensor is `mul_` (3 axes). We hit this assertion `rhs_i >= 0 && lhs_i >= 0` because `view_` output with 2 axes is chosen as the reference tensor.

TODO:

- [x] Select largest output TV from valid reference tensors when selecting pointwise heuristics

### CODE REVIEW

File: test/test_jit_cuda_fuser.py [hunk 1]
```diff
@@ -4107,6 +4107,29 @@ def t(t0, t1, t2):
...
+    def test_pointwise_reference_tensor(self):
+        def t(input1, input2, scalar):
+            _unsafe_view = torch.ops.aten._unsafe_view(input1, [8, 512, 3072])
+            add_ = torch.ops.aten.add_(_unsafe_view, input2)
+            gelu_ = torch.ops.aten.gelu(add_)
+            view_ = torch.ops.aten.view(gelu_, [4096, 3072])
+            mul_ = torch.ops.aten.mul(add_, scalar)
+            return [view_, mul_]
+
+        x = torch.randn(4096, 3072, device="cuda")
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py]: nitpick: Does the original problem repro with a smaller tensor?

**Review by jjsjann123 — COMMENTED** [review]
Looks like it fixes the AOTAutograd failure. Awesome!

File: test/test_jit_cuda_fuser.py (hunk 1 above)
Comment (rdspring1) [reply test/test_jit_cuda_fuser.py]: Yes, I'll use smaller tensors.

**Comment by csarofeen** [comment]
We should either disable the 2D PWise scheduler with view ops or make sure the analysis is based on the reference tensor and propagates through views correctly.

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp [hunk 2]
```diff
@@ -22,6 +22,144 @@ namespace {
...
+
+      if (!areAllMapped(input_tv, output_tv)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+ private:
+  bool hasMinimumSize(TensorView* tv, int num_axes) const {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp:81]: This is new, everything else in this block was moved up from below.

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp [hunk 3]
```diff
@@ -371,134 +517,6 @@ size_t nRootDims(const TensorView* tv) {
...
-      : fusion_(fusion),
-        ca_index_map_(ComputeAtMap(ComputeAtMap::MappingMode::INDEX)) {
-    ca_index_map_.build(fusion);
-    view_tvs_ = scheduler_utils::getViewTVs(fusion);
-  }
-
-  TensorView* findReferenceTensorView() const {
-    auto fusion_outputs = fusion_->outputs();
-    for (auto output_tv : ir_utils::filterByType<TensorView>(fusion_outputs)) {
-      if (isValidReference(output_tv) && !output_tv->isFusionInput()) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp:389]: I think we need to make sure the `!output_tv->isFusionInput()` propagates.

**Review by csarofeen — APPROVED** [review]
Looks strictly better, the only thing that seems odd to me is the check `hasMinimumSize`, is this still required or has this changed since you check the reference TV is consistent with the check on breakpoint? Approving since I'm assuming it's spurious.

**Review by csarofeen — APPROVED** [review]
LGTM discussed this

**Pull request merged**